File: php/nodes/read_m3u_file.php
<?php

function read_m3u_file(array $edges, array $utils = []): callable {
    $utils += ['read_text_file' => fn($p) => file_get_contents($p)];

    return function ($state) use ($edges, $utils): Next {
        if (empty($state['m3u_file_path'])) {
            throw new \RuntimeException('M3U file path is not defined.');
        }

        $content = $utils['read_text_file']($state['m3u_file_path']);
        $urls = array_values(array_filter(
            array_map('trim', explode("\n", $content)),
            fn($line) => $line !== '' && !str_starts_with($line, '#')
        ));

        echo "Found " . count($urls) . " URLs in {$state['m3u_file_path']}\n";
        return new Next($edges['on_success'], array_merge($state, ['urls' => $urls]));
    };
}

File: ruby/nodes/read_m3u_file.rb
# frozen_string_literal: true

require_relative '../lib/ambler'

module ReadM3uFile
  DEFAULT_UTILS = { read_text_file: ->(path) { File.read(path) } }.freeze

  def self.node(edges, utils = DEFAULT_UTILS)
    ->(state) {
      raise 'M3U file path is not defined.' if state[:m3u_file_path].nil?

      content = utils[:read_text_file].call(state[:m3u_file_path])
      urls = content.lines.map(&:strip).reject { |line| line.empty? || line.start_with?('#') }

      puts "Found #{urls.length} URLs in #{state[:m3u_file_path]}"
      Ambler::Next.new(edges[:on_success], state.merge(urls: urls))
    }
  end
end
